### PR TITLE
DRILL-7429: Wrong column order when selecting complex data using Hive storage plugin

### PR DESCRIPTION
**Pull request by ihuzenko**
# [DRILL-7429](https://issues.apache.org/jira/browse/DRILL-7429): Wrong column order when selecting complex data using Hive storage plugin

## Description

Column ordering was wrong when the first selected column had type STRUCT and second had a primitive type. 

## Documentation

No need to document it.

## Testing

Added unit test in TestHiveStructs.

### CODE REVIEW

File: contrib/storage-hive/core/src/test/java/org/apache/drill/exec/hive/complex_types/TestHiveStructs.java [hunk 1]
```diff
@@ -234,6 +240,28 @@ public void primitiveStruct() throws Exception {
...
+    String sql = "SELECT t.str_n0 a, rid b FROM hive.struct_tbl t LIMIT 1";
+    DirectRowSet directRowSet = queryBuilder().sql(sql).rowSet();
+    BatchSchema fields;
+    try {
+      fields = directRowSet.batchSchema();
+    } finally {
+      directRowSet.clear();
+    }
+
+    assertThat(fields.getFieldCount(), is(2));
```
Comment (vvysotskyi) [inline contrib/storage-hive/core/src/test/java/org/apache/drill/exec/hive/complex_types/TestHiveStructs.java:254]: We also have `RowSetComparison` class, which allows comparing expected and actual row sets, including their schema.

Or if you don't want to populate row set with expected data, you may use the `schemaBaseLine()` method from `TestBuilder` to validate query schema only.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java [hunk 2]
```diff
@@ -120,17 +119,7 @@ protected IterOutcome doWork() {
...
-            // been setup during setupNewSchema
-            for (FieldReference fieldReference : complexFieldReferencesList) {
-              MaterializedField field = MaterializedField.create(fieldReference.getAsNamePart().getName(),
-                      UntypedNullHolder.TYPE);
-              container.add(new UntypedNullVector(field, container.getAllocator()));
-            }
-            container.buildSchema(SelectionVectorMode.NONE);
-            wasNone = true;
-            return IterOutcome.OK_NEW_SCHEMA;
+            return IterOutcome.NONE;
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java:122]: We haven't built the schema. I think even for the case when incoming operator returned `NONE` for the first batch, we should also build the schema, in the opposite case, we may break some cases of `limit 0` queries.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java [hunk 3]
```diff
@@ -339,6 +331,14 @@ protected IterOutcome handleNullInput() {
     return IterOutcome.OK_NEW_SCHEMA;
   }
 
+  private boolean isNotSchemalessInput() {
+    RecordBatch incomingBatch = incoming instanceof IteratorValidatorBatchIterator
+        ? ((IteratorValidatorBatchIterator) incoming).getIncoming()
+        : incoming;
```
Comment (vvysotskyi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java:337]: Is it required to ensure that incoming batch after `IteratorValidatorBatchIterator` is `IteratorValidatorBatchIterator`?

Looks like `SchemalessBatch.getSchema()` always returns null, so at least the second check covers this `instance of` check...

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java (hunk 2 above)
Comment (ihuzenko) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java:122]: AFAIK there is ```handleNullInput()``` which should be called for this purpose. Here was workaround for adding late complex types to container and rebuilding schema was necessary. One thing I may be missing here is setting ```wasNone = true;```, so I'll add it.

**Comment by ihuzenko** [comment]
Hello @vvysotskyi , thanks for good review. I've updated the PR, please take a look again.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java (hunk 2 above)
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectRecordBatch.java:122]: Thanks for referring to `handleNullInput()` and explanation.

**Pull request merged**